File: android/app/src/main/java/blog/nextlab/echo/backup/BackupArchive.kt
package blog.nextlab.echo.backup

import blog.nextlab.echo.data.MessageEnvelope
import blog.nextlab.echo.core.model.ConversationId
import blog.nextlab.echo.core.model.MessageContent
import blog.nextlab.echo.core.model.MessageId
import blog.nextlab.echo.core.model.UserId
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * バックアップに何が入り、どう書かれるか。
 *
 * 入るのはメッセージで、メディアは入らない。写真や動画は保管庫から取り直せるが、
 * メッセージは開く鍵が消えたらどこからも戻らない。なので書庫は会話と、
 * **その中のメディアへの参照**（id と鍵）を運び、バイト列は運ばない。すでにサーバーに
 * ある動画で他人の15GBを埋めるのは割に合わないし、それこそ人がバックアップを
 * 切る理由になる。
 *
 * 中身を MessageEnvelope で書くのは、メッセージ自身が通るのと同じ形式だから。
 * 符号化器も復号器もテストも1組で済み、写真が復元を生き延びる理由は送信を生き延びる
 * 理由とまったく同じになる。同じものに形式を2つ作れば必ずずれ、そのずれは
 * 誰かが復元する日にだけ現れる。
 *
 * 版の番号は書いて確認する。新しいビルドが書いたファイルは、半分読まずに拒否する。
 * 会話の部分的な復元は、何も戻らないより悪い（うまくいったように見えるから）。
 */
object BackupArchive {

    const val VERSION = 1

    /** 復号済みのメッセージ1件。書き戻される形。 */
    class Entry(
        val conversationId: ConversationId,
        val messageId: MessageId,
        val senderId: UserId,
        val sentAtMs: Long,
        val content: MessageContent,
    )

    class Parsed(
        val version: Int,
        val owner: UserId,
        val createdAtMs: Long,
        val entries: List<Entry>,
    )

    fun write(owner: UserId, createdAtMs: Long, entries: List<Entry>): String {
        val array = JSONArray()
        for (entry in entries) {
            array.put(
                JSONObject()
                    .put(CONVERSATION, entry.conversationId.value)
                    .put(MESSAGE, entry.messageId.value)
                    .put(SENDER, entry.senderId.value)
                    .put(SENT_AT, entry.sentAtMs)
                    // 封の中身をそのまま。MessageEnvelope を参照。
                    .put(CONTENT, MessageEnvelope.seal(entry.content)),
            )
        }

        return JSONObject()
            .put(VERSION_KEY, VERSION)
            .put(OWNER, owner.value)
            .put(CREATED_AT, createdAtMs)
            .put(MESSAGES, array)
            .toString()
    }

    /**
     * 書庫を読む。このビルドが解せるものでなければ null。
     *
     * null になるのは「JSON でない」「版が無い」「もっと新しいものが書いた」の3つ。
     * 復元する人にとってはどれも同じ意味で、このファイルはここでは使えない。
     */
    fun read(json: String): Parsed? {
        val root = try {
            JSONObject(json)
        } catch (_: JSONException) {
            return null
        }

        val version = root.optInt(VERSION_KEY, 0)
        if (version <= 0 || version > VERSION) return null

        val owner = root.optString(OWNER).takeIf { it.isNotEmpty() } ?: return null
        val messages = root.optJSONArray(MESSAGES) ?: JSONArray()

        val entries = buildList {
            for (i in 0 until messages.length()) {
                val item = messages.optJSONObject(i) ?: continue
                val content = MessageEnvelope.open(item.optString(CONTENT)) ?: continue
                val conversation = item.optString(CONVERSATION).takeIf { it.isNotEmpty() }
                    ?: continue
                val message = item.optString(MESSAGE).takeIf { it.isNotEmpty() } ?: continue
                val sender = item.optString(SENDER).takeIf { it.isNotEmpty() } ?: continue

                add(
                    Entry(
                        conversationId = ConversationId(conversation),
                        messageId = MessageId(message),
                        senderId = UserId(sender),
                        sentAtMs = item.optLong(SENT_AT),
                        content = content,
                    ),
                )
            }
        }

        return Parsed(
            version = version,
            owner = UserId(owner),
            createdAtMs = root.optLong(CREATED_AT),
            entries = entries,
        )
    }

    private const val VERSION_KEY = "v"
    private const val OWNER = "owner"
    private const val CREATED_AT = "at"
    private const val MESSAGES = "messages"
    private const val CONVERSATION = "c"
    private const val MESSAGE = "m"
    private const val SENDER = "s"
    private const val SENT_AT = "t"
    private const val CONTENT = "e"
}

File: ios/RinowaCore/Sources/RinowaCore/BackupArchive.swift
import Foundation

/// バックアップに何が入るか。`BackupArchive.kt` の Swift 側。
///
/// メッセージと、写真に届く鍵。写真のバイト列は入れない。写真は保管庫から取り直せるが、
/// メッセージは開く鍵が消えたらどこからも戻らない。docs/WIRE_FORMATS.md §2.2。
public enum BackupArchive {

    public static let version = 1

    public struct Entry: Equatable {
        public let conversationId: String
        public let messageId: String
        public let senderId: String
        public let sentAtMs: Int64
        public let content: MessageContent

        public init(
            conversationId: String,
            messageId: String,
            senderId: String,
            sentAtMs: Int64,
            content: MessageContent
        ) {
            self.conversationId = conversationId
            self.messageId = messageId
            self.senderId = senderId
            self.sentAtMs = sentAtMs
            self.content = content
        }
    }

    public struct Parsed: Equatable {
        public let version: Int
        public let owner: String
        public let createdAtMs: Int64
        public let entries: [Entry]
    }

    public static func write(owner: String, createdAtMs: Int64, entries: [Entry]) -> String {
        let messages: [[String: Any]] = entries.map { entry in
            [
                "c": entry.conversationId,
                "m": entry.messageId,
                "s": entry.senderId,
                "t": entry.sentAtMs,
                // 封の中身をそのまま。メッセージが存在する2箇所で符号化器を1つにする。
                "e": MessageEnvelope.seal(entry.content)
            ]
        }

        let root: [String: Any] = [
            "v": version,
            "owner": owner,
            "at": createdAtMs,
            "messages": messages
        ]

        guard
            let data = try? JSONSerialization.data(withJSONObject: root),
            let text = String(data: data, encoding: .utf8)
        else {
            return ""
        }
        return text
    }

    /// 書庫を読む。このビルドが解せるものでなければ nil。
    ///
    /// 新しいビルドが書いたファイルは、半分読まずに拒否する。ばらばらに戻ってくる会話は、
    /// 戻ってこない会話より悪い。うまくいったように見えるから。
    public static func read(_ json: String) -> Parsed? {
        guard
            let data = json.data(using: .utf8),
            let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let rawVersion = (root["v"] as? NSNumber)?.intValue,
            rawVersion > 0, rawVersion <= version,
            let owner = root["owner"] as? String, !owner.isEmpty
        else {
            return nil
        }

        let messages = root["messages"] as? [[String: Any]] ?? []
        let entries: [Entry] = messages.compactMap { item in
            guard
                let conversation = item["c"] as? String, !conversation.isEmpty,
                let message = item["m"] as? String, !message.isEmpty,
                let sender = item["s"] as? String, !sender.isEmpty,
                let envelope = item["e"] as? String,
                let content = MessageEnvelope.open(envelope)
            else {
                return nil
            }

            return Entry(
                conversationId: conversation,
                messageId: message,
                senderId: sender,
                sentAtMs: Int64((item["t"] as? NSNumber)?.int64Value ?? 0),
                content: content
            )
        }

        return Parsed(
            version: rawVersion,
            owner: owner,
            createdAtMs: Int64((root["at"] as? NSNumber)?.int64Value ?? 0),
            entries: entries
        )
    }
}
